### ISSUE TITLE
Client oom 3.8-SNAPSHOT

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#9062]
Client oom 3.8-SNAPSHOT  
in shutdown tests 

https://hazelcast-l337.ci.cloudbees.com/view/shutdown/

```
shutdown-set/3.8-SNAPSHOT/2016_10_07-04_47_06/output/HZ/HzClient2HZ/HzClient2HZ.hprof
shutdown-set/3.8-SNAPSHOT/2016_10_06-04_47_06/output/HZ/HzClient1HZ/HzClient1HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient3HZ/HzClient3HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient4HZ/HzClient4HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient2HZ/HzClient2HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient1HZ/HzClient1HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient3HZ/HzClient3HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient4HZ/HzClient4HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient2HZ/HzClient2HZ.hprof
shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient1HZ/HzClient1HZ.hprof
```

http://54.87.52.100/~jenkins/workspace/shutdown-set/3.8-SNAPSHOT/2016_10_07-04_47_06/output/HZ/HzClient2HZ/

http://54.87.52.100/~jenkins/workspace/shutdown-set/3.8-SNAPSHOT/2016_10_06-04_47_06/output/HZ/HzClient1HZ/

http://54.87.52.100/~jenkins/workspace/shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient3HZ/

**Comment by Danny-Hazelcast** [comment]
INFO: hz.client_0 [HZ] [3.8-SNAPSHOT] HazelcastClient 3.8-SNAPSHOT (20161005 - 75806d4) is CLIENT_CONNECTED

**Comment by Danny-Hazelcast** [comment]
http://54.87.52.100/~jenkins/workspace/shutdown-set/3.8-SNAPSHOT/2016_10_07-04_47_06/output/HZ/HzClient2HZ/

<img width="651" alt="screen shot 2016-10-07 at 09 25 56" src="https://cloud.githubusercontent.com/assets/5988678/19180589/a7dcece6-8c6f-11e6-8a2d-ab06ea350dea.png">

**Comment by Danny-Hazelcast** [comment]
I can not reproduce this issue using 3.7.2

**Comment by jerrinot** [comment]
thkns! so it appears to be 3.8-specific? hence I'm moving milestone to 3.8

**Comment by jerrinot** [comment]
@Danny-Hazelcast: what is the test doing? is the a chance it would try to allocate a massive object / array in any given point in time? the heapdump is almost empty yet there is OOM. 

I presume the OOM happened due an allocation request which couldn't be satisfied.

**Comment by Danny-Hazelcast** [comment]
there are in fact 2 test failing with the OOM

shutdown-set
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-set/
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-set/20/console  (3.8-SNAP)
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-set/21/  3.7.2

shutdown-sql
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/42/console

normally shutdown-sql is failing with https://github.com/hazelcast/hazelcast/issues/8931

**Comment by jerrinot** [comment]
@Danny-Hazelcast: When did it start to fail? can you localize the commit?

**Comment by Danny-Hazelcast** [comment]
NFO: hz.client_0 [HZ] [3.8-SNAPSHOT] HazelcastClient 3.8-SNAPSHOT (20161005 - 75806d4) is CLIENT_CONNECTED

**Comment by Danny-Hazelcast** [comment]
client jvm settings,   -Xms200M -Xmx400M

**Comment by Danny-Hazelcast** [comment]
and failed again with 3.7.2 
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-set/22/console
http://54.87.52.100/~jenkins/workspace/shutdown-set/3.8-SNAPSHOT/2016_10_07-11_15_45/

**Comment by Danny-Hazelcast** [comment]
this client side OOME is provoked in a whole range of tests.

https://hazelcast-l337.ci.cloudbees.com/view/shutdown/

```
-rwxr-xr-x 1 jenkins jenkins 8.0M Oct  8 14:01 shutdown-atomic/3.8-SNAPSHOT/2016_10_08-16_57_31/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 8.1M Oct  9 02:08 shutdown-atomic/3.8-SNAPSHOT/2016_10_09-04_54_05/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct 10 01:56 shutdown-atomic/3.8-SNAPSHOT/2016_10_10-04_54_05/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 9.4M Oct  8 02:06 shutdown-data-loss/3.8-SNAPSHOT/2016_10_08-05_00_08/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 9.4M Oct  9 02:30 shutdown-data-loss/3.8-SNAPSHOT/2016_10_09-05_00_11/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 9.2M Oct 10 02:05 shutdown-data-loss/3.8-SNAPSHOT/2016_10_10-05_00_08/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct 10 01:58 shutdown-lock/3.8-SNAPSHOT/2016_10_10-04_56_05/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 8.1M Oct 10 01:58 shutdown-lock/3.8-SNAPSHOT/2016_10_10-04_56_05/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 8.3M Oct  8 01:44 shutdown-map-lock/3.8-SNAPSHOT/2016_10_08-04_15_07/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 8.5M Oct 10 01:17 shutdown-map-lock/3.8-SNAPSHOT/2016_10_10-04_15_07/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct  6 01:52 shutdown-set/3.8-SNAPSHOT/2016_10_06-04_47_06/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.9M Oct  7 02:01 shutdown-set/3.8-SNAPSHOT/2016_10_07-04_47_06/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct  7 08:43 shutdown-set/3.8-SNAPSHOT/2016_10_07-11_15_45/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct  7 08:43 shutdown-set/3.8-SNAPSHOT/2016_10_07-11_15_45/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.8M Oct  8 01:52 shutdown-set/3.8-SNAPSHOT/2016_10_08-04_47_05/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 7.9M Oct  9 01:54 shutdown-set/3.8-SNAPSHOT/2016_10_09-04_47_05/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  11M Oct 10 01:49 shutdown-set/3.8-SNAPSHOT/2016_10_10-04_47_05/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  11M Oct 10 01:49 shutdown-set/3.8-SNAPSHOT/2016_10_10-04_47_05/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  6 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  6 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  6 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  6 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_06-05_00_14/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  19M Oct  7 02:05 shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  7 02:05 shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct  7 02:05 shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  7 02:05 shutdown-sql/3.8-SNAPSHOT/2016_10_07-05_02_10/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct  8 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_08-05_00_07/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins 8.6M Oct  8 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_08-05_00_07/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  8 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_08-05_00_07/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  16M Oct  8 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_08-05_00_07/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct  9 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_09-05_00_07/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  35M Oct  9 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_09-05_00_07/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct  9 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_09-05_00_07/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct  9 02:03 shutdown-sql/3.8-SNAPSHOT/2016_10_09-05_00_07/output/HZ/HzClient4HZ/HzClient4HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct 10 02:04 shutdown-sql/3.8-SNAPSHOT/2016_10_10-05_00_08/output/HZ/HzClient1HZ/HzClient1HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct 10 02:04 shutdown-sql/3.8-SNAPSHOT/2016_10_10-05_00_08/output/HZ/HzClient2HZ/HzClient2HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  14M Oct 10 02:04 shutdown-sql/3.8-SNAPSHOT/2016_10_10-05_00_08/output/HZ/HzClient3HZ/HzClient3HZ.hprof
-rwxr-xr-x 1 jenkins jenkins  15M Oct 10 02:04 shutdown-sql/3.8-SNAPSHOT/2016_10_10-05_00_08/output/HZ/HzClient4HZ/HzClient4HZ.hprof
```

**Comment by jerrinot** [comment]
fixed by #9091
